### PR TITLE
fix(cli): Handle non-unicode characters in plugin logs

### LINKED ISSUE #8799: bug: Handle non-unicode characaters in plugin logs
### Meltano Version

3.5.0

### Python Version

3.10

### Bug scope

CLI (options, error messages, logging, etc.)

### Operating System

Windows 11 - pwsh

### Description

running tap-mssql to target-csv

records with non-ascii characters ("самых"), pipeline runs but the logging throws errors:

```
Arguments: ()
--- Logging error ---
Traceback (most recent call last):
  File "C:\Users\<user>\AppData\Local\Programs\Python\Python310\lib\logging\__init__.py", line 1103, in emit
    stream.write(msg + self.terminator)
  File "C:\Users\<user>\AppData\Local\Programs\Python\Python310\lib\encodings\cp1252.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_table)[0]
UnicodeEncodeError: 'charmap' codec can't encode characters in position 124-126: character maps to <undefined>
Call stack:
  File "C:\Users\<user>\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\<user>\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "c:\users\<user>\.local\bin\meltano.exe\__main__.py", line 7, in <module>
    sys.exit(main())
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\meltano\cli\__init__.py", line 124, in main
    _run_cli()
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\meltano\cli\__init__.py", line 103, in _run_cli
    cli(obj={"project": None})
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\click\core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\meltano\cli\cli.py", line 46, in main
    return super().main(*args, windows_expand_args=False, **kwargs)
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\click\core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\meltano\cli\utils.py", line 662, in invoke
    super().invoke(ctx)  # type: ignore[misc]
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\click\core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\meltano\cli\utils.py", line 711, in invoke
    super().invoke(ctx)
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\click\core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\click\core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\meltano\cli\params.py", line 74, in decorate
    return func(*args, **kwargs)
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\meltano\cli\params.py", line 164, in decorate
    func(project, *args, **kwargs)
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\click\decorators.py", line 33, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\meltano\core\utils\__init__.py", line 73, in wrapper
    return asyncio.run(func(*args, **kwargs))
  File "C:\Users\<user>\AppData\Local\Programs\Python\Python310\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Users\<user>\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 636, in run_until_complete
    self.run_forever()
  File "C:\Users\<user>\AppData\Local\Programs\Python\Python310\lib\asyncio\windows_events.py", line 321, in run_forever
    super().run_forever()
  File "C:\Users\<user>\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 603, in run_forever
    self._run_once()
  File "C:\Users\<user>\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 1909, in _run_once
    handle._run()
  File "C:\Users\<user>\AppData\Local\Programs\Python\Python310\lib\asyncio\events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\meltano\core\logging\utils.py", line 230, in capture_subprocess_output
    if not await _write_line_writer(writer, line):
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\meltano\core\logging\utils.py", line 201, in _write_line_writer
    writer.writeline(line.decode(errors="replace"))
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\meltano\core\logging\output_logger.py", line 271, in writeline
    self.logger.log(self.write_level, line.rstrip(), name=self.name)
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\structlog\stdlib.py", line 237, in log
    return self._proxy_to_logger(LEVEL_TO_NAME[level], event, *args, **kw)
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\structlog\stdlib.py", line 258, in _proxy_to_logger
    return super()._proxy_to_logger(method_name, event=event, **event_kw)
  File "C:\Users\<user>\pipx\venvs\meltano\lib\site-packages\structlog\_base.py", line 215, in _proxy_to_logger
    return getattr(self._logger, method_name)(*args, **kw)
Message: {'consumer': False, 'producer': True, 'string_id': 'tap-mssql--ap', 'cmd_type': 'elb', 'run_id': 'de7cd47c-a1ea-4dcb-93e1-c1503efd5c3b', 'job_name': 'local:tap-mssql--ap-to-target-csv', 'stdio': 'stdout', 'name': 'tap-mssql--ap', 'event': '{"type":"RECORD","stream": ---REDACTED--- entry: "самых" ---REDACTED---   ,"version":1727252247132,"time_extracted":"2024-09-25T08:17:27.166450Z"}', 'level': 'info', 'timestamp': '2024-09-25T08:17:30.369830Z'}
```

### Code

_No response_

### Link to Slack/Linen

_No response_

### PR DESCRIPTION
**Pull request by toroleapinc**
## Summary

On Windows, the default console encoding (e.g. cp1252) cannot represent all Unicode characters. When plugins (e.g. `tap-mssql`) emit log lines containing non-ASCII text such as Cyrillic characters, the standard `logging.StreamHandler` raises `UnicodeEncodeError` and drops the log entry entirely.

## Changes

- Add `SafeStreamHandler` class in `src/meltano/core/logging/utils.py` that extends `logging.StreamHandler` with a fallback `emit()` method using `backslashreplace` error handling
- Update the console handler configuration to use `SafeStreamHandler` instead of `logging.StreamHandler`

This ensures that log output containing non-encodable characters is never lost — instead, unencodable characters are escaped (e.g. `\u0441\u0430\u043c\u044b\u0445`).

Fixes #8799

## Summary by Sourcery

Bug Fixes:
- Prevent UnicodeEncodeError from dropping plugin log lines containing non-ASCII characters in console output.

### CODE REVIEW

**Review by edgarrmondragon — CHANGES_REQUESTED** [review]
Tests?

**Comment by toroleapinc** [comment]
Thanks for the feedback! You're absolutely right - I should add tests for this new SafeStreamHandler. I'll work on writing comprehensive tests that cover:

1. Normal operation (should delegate to super().emit())
2. UnicodeEncodeError handling with backslashreplace fallback
3. Edge cases like streams without encoding attributes
4. Error handling for I/O failures in the fallback path

I'll also incorporate the excellent suggestions from @sourcery-ai about making the encoding detection more robust and preserving the standard logging error handling behavior.

Working on this now and will push an update soon!

File: tests/meltano/core/logging/test_logging_utils.py [hunk 1]
```diff
@@ -197,3 +198,156 @@ def test_disabled_log_level():
...
+            msg="Message with unicode: 中文",
+            args=(),
+            exc_info=None,
+        )
+
+        # Should not raise an exception
+        handler.emit(record)
+
+        # Verify that write was called twice (first failed, second succeeded)
+        assert mock_stream.write.call_count == 2
```
Comment (edgarrmondragon) [inline tests/meltano/core/logging/test_logging_utils.py]: Should the test be calling `.emit()` twice to exercise both side effects?

File: tests/meltano/core/logging/test_logging_utils.py [hunk 2]
```diff
@@ -197,3 +198,156 @@ def test_disabled_log_level():
...
+            msg="Message with unicode: 中文",
+            args=(),
+            exc_info=None,
+        )
+
+        # Should not raise an exception and should handle encoding fallback
+        handler.emit(record)
+
+        # Verify that write was called twice
+        assert mock_stream.write.call_count == 2
```
Comment (edgarrmondragon) [inline tests/meltano/core/logging/test_logging_utils.py]: Ditto

File: tests/meltano/core/logging/test_logging_utils.py [hunk 3]
```diff
@@ -197,3 +198,156 @@ def test_disabled_log_level():
...
```
Comment (edgarrmondragon) [inline tests/meltano/core/logging/test_logging_utils.py]: Ditto

File: tests/meltano/core/logging/test_logging_utils.py (hunk 1 above)
Comment (toroleapinc) [reply tests/meltano/core/logging/test_logging_utils.py]: Good catch — you were right that this wasn't actually working. The root cause was that `super().emit()` already catches all exceptions internally (including `UnicodeEncodeError`) via its own `except Exception` handler, so our fallback path never fired. That's why `write.call_count` was 1 instead of 2.

Fixed by reimplementing the write logic directly instead of delegating to `super().emit()`, so we can catch `UnicodeEncodeError` before the general `Exception` handler. Also added `except RecursionError: raise` and `except Exception: self.handleError(record)` to match the upstream `StreamHandler.emit()` semantics.

With the fix, a single `.emit()` call correctly exercises both side effects — first write fails with `UnicodeEncodeError`, then the fallback write succeeds. Also split the fallback error test to properly enter the unicode fallback path first, and added a separate test for the general exception → `handleError` path.

File: tests/meltano/core/logging/test_logging_utils.py (hunk 2 above)
Comment (toroleapinc) [reply tests/meltano/core/logging/test_logging_utils.py]: Addressed — same fix applies here. See my reply above for the full explanation.

File: tests/meltano/core/logging/test_logging_utils.py (hunk 3 above)
Comment (toroleapinc) [reply tests/meltano/core/logging/test_logging_utils.py]: Addressed — same fix applies here.

**Comment by toroleapinc** [comment]
Hey Edgar, thanks for the earlier review! Just pushed a test for that RecursionError branch — codecov should be passing now. The only failing job is a Windows SQLite flake that's been haunting us for 3 days, completely unrelated to the code. Anything else we need to move this forward?

**Comment by toroleapinc** [comment]
Hey @edgarrmondragon — circling back on this one. The test feedback should be addressed now (single `.emit()` call exercises both paths after I stopped delegating to `super().emit()`). Let me know if anything else needs changing here.

**Review by edgarrmondragon — APPROVED** [review]

**Review by edgarrmondragon — APPROVED** [review]
Thanks @toroleapinc!

**Comment by toroleapinc** [comment]
@edgarrmondragon — anything else needed here, or good to land?

**Comment by edgarrmondragon** [comment]
> @edgarrmondragon — anything else needed here, or good to land?

Good to land. Thanks @toroleapinc!

**Comment by edgarrmondragon** [comment]
@all-contributors please add @toroleapinc for code
